akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2HighlevelServerSpec.scala

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2020-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.impl.engine.http2
+
+import akka.http.impl.engine.ws.ByteStringSinkProbe
+import akka.http.impl.util.AkkaSpecWithMaterializer
+import akka.http.impl.util.ExampleHttpContexts
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model.AttributeKey
+import akka.http.scaladsl.model.ContentTypes
+import akka.http.scaladsl.model.HttpEntity
+import akka.http.scaladsl.model.HttpHeader
+import akka.http.scaladsl.model.HttpMethod
+import akka.http.scaladsl.model.HttpMethods
+import akka.http.scaladsl.model.HttpRequest
+import akka.http.scaladsl.model.HttpResponse
+import akka.http.scaladsl.model.RequestResponseAssociation
+import akka.http.scaladsl.model.StatusCode
+import akka.http.scaladsl.model.StatusCodes
+import akka.http.scaladsl.model.Uri
+import akka.http.scaladsl.settings.ClientConnectionSettings
+import akka.http.scaladsl.settings.ServerSettings
+import akka.stream.scaladsl.Sink
+import akka.stream.scaladsl.Source
+import akka.stream.testkit.TestPublisher
+import akka.stream.testkit.TestSubscriber
+import akka.stream.testkit.Utils.TE
+import akka.testkit.TestProbe
+import akka.util.ByteString
+import org.scalatest.concurrent.ScalaFutures
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.concurrent.Promise
+
+class Http2HighlevelServerSpec extends AkkaSpecWithMaterializer(
+  """akka.http.server.remote-address-header = on
+     akka.http.server.http2.log-frames = on
+     akka.http.server.log-unencrypted-network-bytes = 100
+     akka.http.server.enable-http2 = on
+     akka.http.client.http2.log-frames = on
+     akka.http.client.log-unencrypted-network-bytes = 100
+     akka.actor.serialize-messages = false
+  """) with ScalaFutures {
+
+  case class RequestId(id: String) extends RequestResponseAssociation
+  val requestIdAttr = AttributeKey[RequestId]("requestId")
+
+  "A HTTP 2 server" should {
+
+    "return internal error when handler throws" in new TestSetup {
+      sendClientRequest()
+      val serverRequest = expectServerRequest()
+      serverRequest.promise.failure(TE("boom"))
+      val response = expectClientResponse()
+      response.status should be(StatusCodes.InternalServerError)
+
+      // stream not torn down so we can send another one
+      sendClientRequest()
+      val serverRequest2 = expectServerRequest()
+      serverRequest2.promise.success(HttpResponse())
+      val response2 = expectClientResponse()
+      response2.status should be(StatusCodes.OK)
+
+    }
+  }
+
+  case class ServerRequest(request: HttpRequest, promise: Promise[HttpResponse]) {
```

Comment:
Below is mostly copy pasta from Http2ClientServerSpec but that one has log capturing fail on errors logged so had to introduce new test file for this case.